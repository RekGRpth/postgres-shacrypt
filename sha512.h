#define SHA512_SALT_PREFIX "$6$"
#define SHA512_ROUNDS_PREFIX "rounds="

/* Structure to save state of computation between the single steps.  */
struct sha512_ctx
{
  uint64_t H[8];

  uint64_t total[2];
  uint64_t buflen;
  char buffer[256];	/* NB: always correctly aligned for uint64_t.  */
};

char *sha512_crypt_r (const char *key, const char *salt, char *buffer, int buflen);

char *sha512_crypt (const char *key, const char *salt);

