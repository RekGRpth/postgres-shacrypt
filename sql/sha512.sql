select sha512_crypt('Hello world!', '$6$saltstring');
select sha512_crypt('Hello world!', '$6$rounds=10000$saltstringsaltstring');
select sha512_crypt('This is just a test', '$6$rounds=5000$toolongsaltstring');
select sha512_crypt('a very much longer text to encrypt.  This one even stretches over morethan one line.', '$6$rounds=1400$anotherlongsaltstring');
select sha512_crypt('we have a short salt string but not a short password', '$6$rounds=77777$short');
select sha512_crypt('a short string', '$6$rounds=123456$asaltof16chars..');
select sha512_crypt('the minimum number is still observed', '$6$rounds=10$roundstoolow');
