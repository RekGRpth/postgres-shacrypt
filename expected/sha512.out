select sha512_crypt('Hello world!', '$6$saltstring');
                                             sha512_crypt                                             
------------------------------------------------------------------------------------------------------
 $6$saltstring$svn8UoSVapNtMuq1ukKS4tPQd8iKwSMHWjl/O817G3uBnIFNjnQJuesI68u4OTLiBFdcbYEdFCoEOfaS35inz1
(1 row)

select sha512_crypt('Hello world!', '$6$rounds=10000$saltstringsaltstring');
                                                      sha512_crypt                                                       
-------------------------------------------------------------------------------------------------------------------------
 $6$rounds=10000$saltstringsaltst$OW1/O6BYHV6BcXZu8QVeXbDWra3Oeqh0sbHbbMCVNSnCM/UrjmM0Dp8vOuZeHBy/YTBmSK6H9qs/y3RnOaw5v.
(1 row)

select sha512_crypt('This is just a test', '$6$rounds=5000$toolongsaltstring');
                                                      sha512_crypt                                                      
------------------------------------------------------------------------------------------------------------------------
 $6$rounds=5000$toolongsaltstrin$lQ8jolhgVRVhY4b5pZKaysCLi0QBxGoNeKQzQ3glMhwllF7oGDZxUhx1yxdYcz/e1JSbq3y6JMxxl8audkUEm0
(1 row)

select sha512_crypt('a very much longer text to encrypt.  This one even stretches over morethan one line.', '$6$rounds=1400$anotherlongsaltstring');
                                                      sha512_crypt                                                      
------------------------------------------------------------------------------------------------------------------------
 $6$rounds=1400$anotherlongsalts$POfYwTEok97VWcjxIiSOjiykti.o/pQs.wPvMxQ6Fm7I6IoYN3CmLs66x9t0oSwbtEW7o7UmJEiDwGqd8p4ur1
(1 row)

select sha512_crypt('we have a short salt string but not a short password', '$6$rounds=77777$short');
                                                 sha512_crypt                                                 
--------------------------------------------------------------------------------------------------------------
 $6$rounds=77777$short$WuQyW2YR.hBNpjjRhpYD/ifIw05xdfeEyQoMxIXbkvr0gge1a1x3yRULJ5CCaUeOxFmtlcGZelFl5CxtgfiAc0
(1 row)

select sha512_crypt('a short string', '$6$rounds=123456$asaltof16chars..');
                                                       sha512_crypt                                                       
--------------------------------------------------------------------------------------------------------------------------
 $6$rounds=123456$asaltof16chars..$BtCwjqMJGx5hrJhZywWvt0RLE8uZ4oPwcelCjmw2kSYu.Ec6ycULevoBK25fs2xXgMNrCzIMVcgEJAstJeonj1
(1 row)

select sha512_crypt('the minimum number is still observed', '$6$rounds=10$roundstoolow');
                                                    sha512_crypt                                                    
--------------------------------------------------------------------------------------------------------------------
 $6$rounds=1000$roundstoolow$kUMsbe306n21p9R.FRkW3IGn.S9NPN0x50YhH1xhLsPuWGsUSklZt58jaTfF4ZEQpyUNGc0dqbpBYYBaHHrsX.
(1 row)

